modules/layers/src/text-layer/text-layer.js

```diff
@@ -266,63 +304,121 @@ export default class TextLayer extends CompositeLayer {
-          getPosition: updateTriggers.getPosition,
-          getAngle: updateTriggers.getAngle,
-          getColor: updateTriggers.getColor,
-          getSize: updateTriggers.getSize,
-          getPixelOffset: updateTriggers.getPixelOffset,
-          getIconOffsets: {
-            getText: updateTriggers.getText,
-            getTextAnchor: updateTriggers.getTextAnchor,
-            getAlignmentBaseline: updateTriggers.getAlignmentBaseline,
-            styleVersion
+    const CharacgtersLayerClass = this.getSubLayerClass('characters', MultiIconLayer);
+    const BackgroundLayerClass = this.getSubLayerClass('background', TextBackgroundLayer);
+
+    return [
+      background &&
+        new BackgroundLayerClass(
+          {
+            // background props
+            getFillColor: getBackgroundColor,
+            getLineColor: getBorderColor,
+            getLineWidth: getBorderWidth,
+            padding: backgroundPadding,
+
+            // props shared with characters layer
+            getPosition,
+            getSize,
+            getAngle,
+            getPixelOffset,
+            billboard,
+            sizeScale: sizeScale / this.state.fontAtlasManager.props.fontSize,
+            sizeUnits,
+            sizeMinPixels,
+            sizeMaxPixels,
+
+            transitions: transitions && {
+              getPosition: transitions.getPosition,
+              getAngle: transitions.getAngle,
+              getSize: transitions.getSize,
+              getFillColor: transitions.getBackgroundColor,
+              getLineColor: transitions.getBorderColor,
+              getLineWidth: transitions.getBorderWidth,
+              getPixelOffset: transitions.getPixelOffset
+            }
+          },
+          this.getSubLayerProps({
+            id: 'background',
+            updateTriggers: {
+              getPosition: updateTriggers.getPosition,
+              getAngle: updateTriggers.getAngle,
+              getSize: updateTriggers.getSize,
+              getFillColor: updateTriggers.getBackgroundColor,
+              getLineColor: updateTriggers.getBorderColor,
+              getLineWidth: updateTriggers.getBorderWidth,
+              getPixelOffset: updateTriggers.getPixelOffset,
+              getBoundingRect: {
+                getText: updateTriggers.getText,
+                getTextAnchor: updateTriggers.getTextAnchor,
+                getAlignmentBaseline: updateTriggers.getAlignmentBaseline,
+                styleVersion
+              }
+            }
+          }),
+          {
+            data: data.attributes
+              ? {length: data.length, attributes: data.attributes.background || {}}
+              : data,
+            _dataDiff,
+            // Maintain the same background behavior as <=8.3. Remove in v9?
+            autoHighlight: false,
+            getBoundingRect: this.getBoundingRect.bind(this)
+          }
+        ),
+      new CharacgtersLayerClass(
+        {
+          sdf: fontSettings.sdf,
```

Comment:
400 lines, half of it props copying.

Reply:
We decided to use explicit prop forwarding for all core layers a long time ago. `TileLayer` is the only exception and we are dealing with all sorts of issues.